taichi/backends/opengl/opengl_api.cpp

```diff
@@ -232,41 +227,58 @@ void initialize_opengl() {
   }
 }
 
-GLProgram *compile_glsl_program(std::string source) {
-  GLShader shader(source);
-  GLProgram *program = new GLProgram(shader);
-  program->link();
-  return program;
+CompiledGLSL::CompiledGLSL(std::string source)
```

Comment:
`const std::string&`, otherwise we pay one extra copy (which could be huge because this is the actual source code)

Reply:
Sounds reasonable. But I thought std::string was copy-on-write? Anyway, will do these once I have time.

Reply:
No it's not... And even if some compiler has a COW implementation (no major compiler does this), it's non standard so we cannot rely on it. Anyway, either change to `const std::string&` or `std::move(source_code)` should be good.